### ISSUE TITLE
Automatic [de]serialization

### ISSUE THREAD

**Issue by michaelklishin** [ruby-amqp/bunny#185]
It may be nice to steal a page from Hutch/etc's book and provide a way to serialize
and deserialize published messages and delivered bodies automatically. We can go
as far as having our own middleware (à la Rack or Faraday or Ring).

**Comment by camelpunch** [comment]
Link to book?

**Comment by michaelklishin** [comment]
@camelpunch [Hutch is a project from GoCardless](https://github.com/gocardless/hutch)

**Comment by stefansedich** [comment]
I notice this has been sitting for a while, has there been any more discussion on this one behind the scenes? it is something I was wanting today so what better what to get what I want than to implement it.

**Comment by michaelklishin** [comment]
It needs to be opt-in, that's about as far as my opinion about this go. 

On 7 October 2016 at 10:31:48, Stefan Sedich (notifications@github.com) wrote:
> I notice this has been sitting for a while, has there been any more discussion on this one  
> behind the scenes? it is something I was wanting today so what better what to get what I  
> want than to implement it.
>  
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/ruby-amqp/bunny/issues/185#issuecomment-252172352

--  
MK  

Staff Software Engineer, Pivotal/RabbitMQ

**Comment by stefansedich** [comment]
Agreed, will have a think about it and see what I can come up with!

**Comment by tadman** [comment]
This would be nice especially if it was keyed to the Content-Type information provided as part of the message.

**Comment by mlh758** [comment]
Just came across this, I'd like to chime in on the "needs to be opt in" side. I have a couple subscribers that expect a string of bytes to do some initial processing, then deserialize the important bits.
